## Table of Contents:
* [1.0 Introduction to Python3](#intro-python)
    - [1.0.0 What is Python?](#what-is)
    - [Python](#what-is-python)
    - [Jupyter Notebook](#what-is-jn)
    
* [1.1 Data Types in Python](#data-types)
    - [A first look at Python Data Types](#fldt)
    - [Checking the `type` of a value](#check-type)
    - [1.1.1 Numbers `int` and `floats`](#ints-floats)
     - [Basic Arithmetic](#arithmetic)
     - [The Modulo Operator (`%`)](#modulo)
     - [BIDMAS (PEMDAS)](#bidmas)
     - [Variable Assignments](#variable-assign)
     - [Printing](#printing)
    - [1.1.1 Strings, arrays and tuples](#strings-arrays-tuples)
     - [Indexing](#indexing)
     - [Strings](#strings)
     - [Formatting Strings](#formatting-strings)
      - [Method 1: *f*-strings](#f-strings)
      - [Method 2: `.format()` method](#format-method)

# 1.0 Introduction to Python3 <a class="anchor" id="intro-python"></a>

In the following section, we introduce the concepts of basic python data structures, operations, conditionals and loops. This should get you up to speed on the basic methods which we will often use in the rest of the textbook.

## 1.0.0 What is Python? <a class="anchor" id="what-is"></a>

### Python <a class="anchor" id="what-is-python"></a> 
Python is a **high-level**, **dynamically-typed**, **object-oriented programming** (OOP) language. There's a lot to deconstruct here, so let's define precisely what these terms mean. This will allow us to weigh some pros and cons of Python as a programming language and why it should be your first language...

- **high-level** - a high-level programming language is one that is highly abstracted from machine code. Key data structures in high-level languages include variables, arrays, objects and boolean expressions (`true`/`false`). Conversely, low-level languages are ones that are closer to binary machine code. Low-level languages are commonly more efficient that high-level languages, but we trade that efficiency for ease of readbility, learning and understanding.

- **dynamically-typed** - dynamically-typed languages are ones where the developer does not need declare what **type** a variable should have upon definition. The Python compiler (at runtime) will figure out what certain variables should be (`int`, `array`...). This section will dive deeply into the different data types present in Python. This is a bit of a catch to dynamic-typing, because while it makes Python easier to use for beginners, the type of a variable can sometimes change during the course of a calculation (for example, an `array` can become a `tuple`) which can be a nightmare to debug, however, we will find fixes for these problems throughout the course. 

- **object-oriented** - an object-oriented programming (OOP) language is one that is based on the concept of *objects*. An object is a data structure which contains data. Many variable types that appear in Python (integers, arrays, matrices, dictionaries) are objects. Common OOPs include: Python, C++, C#, Java, JavaScript, MATLAB and many more...

### Jupyter Notebook <a class="anchor" id="what-is-jn"></a>
Jupyter Notebook is the tool that these notes are written in. So we're already off to a great start if you managed to get these running. [Jupyter Notebook](https://jupyter.org/) is a web-based interactive development environment for notebooks, code and data. It's an excellent tool for integrating code snippets into blocks of markdown (the text you are reading now). Markdown is a stripped down, light-weight *markup* langage (same markup as in HTML - HyperText Markup Language). It's just an easier way to write HTML.

Nice, so we have a better understanding what Python is, however, some of these terms have been loosely-defined, but it it sufficient for us to start using Python. If you would like better definitions for some of these terms, this is an opportunity to perhaps search for some research... (SHOW SOME RESOURCES) 

Now that we have a better underst, we can see that Python is a good mix of all the best things we need for programming. It is human-readable (high-level), extremely versatile (from OOP) and easy to write (dynamically-typed). Don't worry if you don't understand this definition... it's nice to have and may serve you well later on in your career as a python developer or if you want to learn other languages in the future. 

Next we shall dive into some basics of python and introduce data types in python. 

# 1.1 Data Types in Python
### A first look at Python Data Types <a class="anchor" id="fldt"></a>
Now that we have defined what python is, it's time to focus on what we can do with it. Python3 comes pre-packaged with the following data-types: 

- `int` - Integers 
- `float` - Floating point numbers 
- `str` - Strings 
- `list` - Lists 
- `dict` - Dictionaries unordered `key`: `value` pairs (e.g. `{'name': 'Spongebob', 'age': 24}`)
- `tuples` - Tuples are immutable lists of elements (e.g. `(1,2,3,"Patrick")`)
- `set` - Sets are lists of unique elements (e.g. `{1,2,34,'Smith Boy'}`)
- `bool` - Booleans are conditional values that are `True` or `False`

| Name                 | Type     | Description                                                         |
| :------------------- | :------: | :------------------------------------------------------------------ |
| Integers             | `int`    | Whole numbers (e.g. `3`, `200`, `87`)               |
| Floating Point Number| `float`  | Numbers with decimal point (e.g. `3.14`, `2.718`, `1.414`)          |
| String               | `str`    | Ordered sequence of characters (e.g. `hello`, `200`, `How are you?`)|
| List                 | `list`   | Ordered sequence of objects (e.g. `[1,2,3,4]`, `["Spongebob", 24]`) |
| Dictionary           | `dict`   | Unordered key value pairs (e.g. `{'name': 'Spongebob', 'age': 24}`) |
| Tuples               | `tuples` | Ordered immutable sequence of objects (e.g. `(1,2,3,"Patrick")`)    |
| Sets                 | `set`    | Unordered collections of objects (e.g. `{1,2,34,'Smith Boy'}`)      |
| Booleans             | `bool`   | Logical values indicating `True` or `False`                         |


We will go through each of these data structures and explain what they are and how they can be manipulated.


### Checking the `type` of a value <a class="anchor" id="check-type"></a>
We can make use of the `type` function to check what the type an objectos is. Have a play around with the code below and enter the different data types above to confirm if the examples match the data given:

In [1]:
type(9)

int

## 1.1.1 Numbers (`int` and `float`) <a class="anchor" id="ints-floats"></a>
In Python, an `int` is just a whole number, as we saw earlier. A `float` is a number with a decimal point. One quick point of note is that in python `2` is an `int` and `2.0` is a `float` (even though normally we would class both as being the integers, python makes a distinction). 

### Basic Aritmetic <a class="anchor" id="arithmetic"></a>
In this section we will basically use python as a calculator. We can perform the following basic mathematical operations in Python: 

In [2]:
# Addition
2+1

3

In [3]:
# Subtraction
31-2

29

In [4]:
# Multiplication
21*3

63

In [5]:
# Division
24/6

4.0

Note that the division returns a number with a decimal point which is a `float` although we initially used `int`s to perform the calculation, this will always be the case. Later, we will consider the **precision** of floating point numbers, however, this is not so important at the moment.

Powers have an interesting syntax in python. We use two asterisks (`**`) to denote taking a number to a power, for examples $4^2 = 16$ and whatever $3.14^3$ is

In [6]:
# Powers
4**2 # =16
3.14**3

30.959144000000002

### The Modulo Operator (`%`) <a class="anchor" id="modulo"></a>
Sometimes, we need to calculate the remainder of a number with respect to another. Suppose that we need to check whether the `int` $26$ is even, we can use the modulo operator to ascertain if it's even or odd

In [7]:
# Remainders
26%2

0

In the example above, $26$ has `0` remainder with respect to division by $2$. This means the $26$ is even. On the other hand, if we did the same calculation but with $31$, we would get a remainder of `1`, thus `31` is odd. 

### BIDMAS (PEMDAS) <a class="anchor" id="bidmas"></a>
Does python obey the rules of BIDMAS (PEMDAS)? Let's find out...

In [8]:
2 + 3 * 10 + 3

35

So here, we see python performs the multiplication first, `(3*10)` then the additions `(2 + 30 + 3)`. Which is `35`. But what if we wanted to perform the additions first, we need to use brackets/parentheses

In [9]:
(2+3)*(10+3)

65

Of course, this still obeys BIDMAS, we perform all calculations in the brackets first and then performs the product of the two brackets. It never hurts to use brackets in any numerical calculations, it is possible for it to be overkill, however it doesn't hurt to be safe.

### Variable Assignments <a class="anchor" id="variable-assign"></a>
We can assign all data types to variables, it's best to introduce this as early as possible. Variable declarations in Python are simple compared to other languages. Suppose, we wanted to assign the number `21` to the variable `num_1`. Note the `=` sign here refers to *assignment* not *equality*. Just like we do so in normal homework problems, we assign the variable $x=2$, it's not that the symbol $x$ is the same as $2$. Let's see why variable assignments are useful: 

In [10]:
num_1 = 21
num_2 = 3
num_3 = num_1*num_2


Here, we assigned two variables `num_1` and `num_2`, and used that to calculate the product of those two variables, which was assigned to the the new variables `num_3`. `num_3` is equal to `63`. We can then call `num_3` in further calculations.

### Printing <a class="anchor" id="printing"></a>
Another important tool in our arsenal is printing. It is possible for us to print all variable types to the console by using the `print` function, like so:

In [11]:
print(num_3)

63


The real value of the print function is seen in more further examples. Printing proves extremely useful in debugging code and checking if operations on variable assignments work as expected. More on debugging later.

## 1.1.2 Strings, arrays and tuples
### Indexing <a class="anchor" id="indexing">

Strings (type `str`) or string types are ordered sequences of characters in python. We create strings in python in two ways: 
- Single quotes: `'Hello World!'`
- Double quotes: `"Hello World!"`

Both examples are strings of the same type. From the table above, we see that strings are **ordered sequences**, which means that they permit the use of *indexing* and *slicing*. Consider the string `'hello'`, the table below shows a labelling of the characters of the string, and what index each character corresponds to in the string


| Characters           | h    | e    | l    |l     |o     |
| :------------------- | :---:| :---:| :---:| :---:| :---:|
| Index                | 0    | 1    | 2    | 3    | 4    |
| Reverse Index        | 0    | -4   | -3   | -2   | -1   |

An *index* in an ordered sequence (such as `str`, `list`,`tuple`) of elements, refers to the position of an element in a the sequence, for example

In [12]:
string = "hello"
print(string[0])

h


In words, *the zeroth element of the `string` is* the letter `h`. Python, as with most other languages, uses zero referencing, the first element is at index position `0`, the second at index position `1` and so on. Suppose that we wanted to reference from the other end of the string, Python also allows for this by using negative indices. The final character of the string is at index position `-1`, the penultimate is at index position `-2` and so on...  

This type of indexing also works for other types, but let's focus on strings for the time being.

### Strings <a class="anchor" id="strings"></a>
Earlier, it was mentioned that we can use strings with single and double quotes, however, just a few important rules to remember, the quotes that enclose a string, should be of the same type, for example, this `"Hello World'` will throw an error. Furthermore, if you use single quotes, you have to be careful of apostrophes, for example, `'I'm going to the park'` doesn't translate to *"I'm going to the park"*. The apostrophe in 'I'm' is read as a string literal which encloses the single-quote at the start. We can circumvent this in two ways: 

 - 1. Use single quotes and use `\'` to use the apostrophe - `'I\'m going to the park'`,
 - 2. Forego, the use of single quotes and just use double quotes - `"I'm going to the park"`
 
Neither is better or worse than the other, just make sure to use a back-slash, when trying to format apostrophes in case 1. Let's print some strings now. Suppose we wanted to print `"Hello World!"`, we would simply pass that string into the `print` function like so

In [13]:
print("Hello World")

Hello World


Easy enough. Let's try something a bit longer... 

In [14]:
print('Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut dictum urna at tellus tristique blandit. Etiam pharetra est et enim commodo rhoncus. Phasellus eu laoreet quam. Phasellus in porttitor enim. Donec porta libero at dui accumsan convallis. Suspendisse potenti. Sed fermentum massa dui, sit amet sollicitudin tellus dignissim ac. Phasellus nec venenatis arcu. Integer neque magna, rhoncus sit amet blandit ut, consectetur at nibh. Nunc tristique blandit leo. Ut eu leo nec tortor finibus sollicitudin feugiat ac nibh.')

Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut dictum urna at tellus tristique blandit. Etiam pharetra est et enim commodo rhoncus. Phasellus eu laoreet quam. Phasellus in porttitor enim. Donec porta libero at dui accumsan convallis. Suspendisse potenti. Sed fermentum massa dui, sit amet sollicitudin tellus dignissim ac. Phasellus nec venenatis arcu. Integer neque magna, rhoncus sit amet blandit ut, consectetur at nibh. Nunc tristique blandit leo. Ut eu leo nec tortor finibus sollicitudin feugiat ac nibh.


This is okay, but the string cuts off until the end of the Jupyter viewport **{OR WHATEVER IT'S CALLED}**, what if we wanted to choose the cutoff to a newline ourselves. For this we use an **escape sequence** (`\n`). Escape sequences allow us to define a newline like this

In [15]:
print("Lorem ipsum dolor sit amet, \nconsectetur adipiscing elit. Ut dictum \nurna at tellus tristique blandit. Etiam pharetra \nest et enim commodo rhoncus. Phasellus \neu laoreet quam. Phasellus in porttitor enim. \nDonec porta libero at dui accumsan convallis. \nSuspendisse potenti. Sed fermentum massa dui, sit \namet sollicitudin tellus dignissim ac. Phasellus nec \nvenenatis arcu. Integer neque magna, rhoncus sit \namet blandit ut, consectetur at nibh. Nunc tristique \nblandit leo. Ut eu leo nec tortor finibus \nsollicitudin feugiat ac nibh.")


Lorem ipsum dolor sit amet, 
consectetur adipiscing elit. Ut dictum 
urna at tellus tristique blandit. Etiam pharetra 
est et enim commodo rhoncus. Phasellus 
eu laoreet quam. Phasellus in porttitor enim. 
Donec porta libero at dui accumsan convallis. 
Suspendisse potenti. Sed fermentum massa dui, sit 
amet sollicitudin tellus dignissim ac. Phasellus nec 
venenatis arcu. Integer neque magna, rhoncus sit 
amet blandit ut, consectetur at nibh. Nunc tristique 
blandit leo. Ut eu leo nec tortor finibus 
sollicitudin feugiat ac nibh.


Despite the fact that the formatting is not excellent, it is another option for us in terms of formatting printed strings. 

Another escape sequence is the `tab` (`\t`), this will force 4 extra spaces: 

In [16]:
print('Hello \t World')

Hello 	 World


### Formatting strings <a class="anchor" id="formatting-strings"></a>
Let's print out some results of some calculations to see how useful strings can be. Suppose that we want to calculate the force `F` required to accelerate a particle of mass `m = 10` by `a = 9.81`. Of course the force will be `F = m * a` from Newton's Second Law.

In [17]:
m = 10
a = 9.81

F = m*a
print(F)

98.10000000000001


With no context, it's hard to tell what this result is printing. Especially if you're using an IDE and not Jupyter. Also it's generally better practice to using some more injective string printing. Will the following work?

In [18]:
print('The force required is F N')

The force required is F N


Run the code snippet to see. If you did, you will see that that did not work. A simple syntax is to use `+` literal to literally add that variable to the string

In [19]:
print('The force required is ' + str(F) + ' N')

The force required is 98.10000000000001 N


Here we converted the `F` into a string and then concatenated the strings together. This is still a bit cluncky and there are still slicker ways to do this.

#### Method 1: *f*-strings <a class="anchor" id="f-strings"></a>
The simplest way to do print strings with variables is using an *f*-strings. This is simple enough, we just declare an `f` before the string like so:

In [20]:
result = f"The force required is {F} N"
print(result)

The force required is 98.10000000000001 N


The benefit of this method is that the *f*-strings are 'injective' and simple to use. It created a template for us to just fill in as we please. We can also inject several variables into the string like this

In [21]:
result = f'The force required to accelerate a particle of mass, {m}kg by {a}ms-2 is {F}N'
print(result)

The force required to accelerate a particle of mass, 10kg by 9.81ms-2 is 98.10000000000001N


**IMPORTANT NOTE**: *f*-strings are only supported in Python 3 onwards, so if you use Python 2, you'll most certainly get errors. 

#### Method 2: The `.format()` method <a class="anchor" id="format-method"></a>
Another method of formatting strings is the `.format()` method. This is a method on strings which allows us to inject some variables into the string template. As before, we have

In [22]:
result = "The force required is {} N".format(F)
print(result)

The force required is 98.10000000000001 N


But what if want to inject several variables into the string? The `.format()` string also supports this with an interesting syntax

In [23]:
result = "The force required is {} N for a mass {}kg with to accelerate by {}ms-2".format(F, m, a)
print(result)

The force required is 98.10000000000001 N for a mass 10kg with to accelerate by 9.81ms-2


So, this is cool, however, it can be confusing because we have to remember the order in which the variables appear. We can fix this easily by storing out value in variables

In [24]:
result = "The force required is {force} N for a mass {mass}kg with to accelerate by {acc}ms-2".format(force=F, mass=m, acc=a)
print(result)

The force required is 98.10000000000001 N for a mass 10kg with to accelerate by 9.81ms-2


The cool thing about this is that we can call the variables more than once:

In [25]:
result = "The force required is {force} N for a mass {force}kg with to accelerate by {force}ms-2".format(force=F, mass=m, acc=a)
print(result)

The force required is 98.10000000000001 N for a mass 98.10000000000001kg with to accelerate by 98.10000000000001ms-2


or in any order we like:

In [26]:
result = "The force required is {mass} N for a mass {acc}kg with to accelerate by {force}ms-2".format(force=F, mass=m, acc=a)
print(result)

The force required is 10 N for a mass 9.81kg with to accelerate by 98.10000000000001ms-2


Of course, these last two examples are not mathematically correct, however, it's just a demonstration of the power of the `.format()` method. We can also set the precision of the results appearing within the string. We can see that `F = 98.10000000000001` has too much precision. Precision formatting allows the following options in `{value:width.precision f}` syntax. For example, if we wanted to show the force to `2` decimal places, we would simply insert the following; `{force:1:2f}`

In [27]:
result = "The force required is {force:1.2f} N for a mass {mass}kg with to accelerate by {acc}ms-2".format(force=F, mass=m, acc=a)
print(result)

The force required is 98.10 N for a mass 10kg with to accelerate by 9.81ms-2


I mentioned earlier that the formatting allows for a `width` option, this is not very useful, but this is what it does

In [28]:
result = "The force required is {force:10.2f} N for a mass {mass}kg with to accelerate by {acc}ms-2".format(force=F, mass=m, acc=a)
print(result)

The force required is      98.10 N for a mass 10kg with to accelerate by 9.81ms-2


So a `width = 10` just adds 10 spaces in front of the value being formatted.

## 1.1.2 Lists
The next data type we will focus on is the `list`. Lists are ordered sequences of elements. List elements are contained within square brackets (`[]`) and are separated by commas (`,`), e.g. `[1,2,3,4, "spongebob"]`. As with strings, lists support indexing and negative indexing and there are a multitude of useful methods pre-built into Python. Another important aspect of lists is that lists can contain other lists. To initialise the list, we do so

In [7]:
my_list = [1,2,3,4,5,'Mr. Krabs']
my_list

[1, 2, 3, 4, 5, 'Mr. Krabs']

*Check the type of the list above*

As with strings, lists support indexing and slicing. We did not mention slicing before as the syntax can be a bit confusing. We can also use slicing on string. To find the first index of a list, we use the `0` index:

In [8]:
my_list[0]

1

Remember that we can get the final element of the list using negative indexing

In [9]:
# Write some code here to return 'Mr. Krabs'

As you can see, the syntax for list and string referencing is identical. There isn't much else to cover here.

#### Slicing
Now we move on to something a bit more involved *slicing*. Slicing is a clever way to use a syntax similar to referencing to get particular subsets of an ordered sequence of elements (strings, lists etc.). Let's use slicing to return the list starting from the third element (at index position `2`)

In [10]:
my_list[2:]

[3, 4, 5, 'Mr. Krabs']

Python recognises the colon (':') as some extra options to the calculation. We won't get into the details of what it is and just accept it for what it does. If you put a colon after a single `index` in the square brackets, it will return all the elements including and after the `index`. 

- We can think of it in the following way: `my_list[starting_here:]`

But we can do more with this syntax... 

If we put the colon before an `index`, like this: 

In [47]:
my_list[:2]

[1, 2]

Then we get everything *up to but not including* that `index`. So we supplied the index `2`, so a list containing the `0`th and `1`st elements was returned. 

-  We can think of this in the following way: `my_list[:stop_here]`

What if we wanted every element starting from the `2` index up to but not including the `5` index. Well we can do that by supplying these indices before and after the colon

In [48]:
my_list[2:5]

[3, 4, 5]

This returned a list of the third, fourth and fifth element within the list. 

Lastly, we can also return every `n`th element in the list. This is done using the following syntax

- We can think of this in the following way: `my_list[::step-size]`

In [49]:
my_list[::3]

[1, 4]

Let's use a slightly longer list for this example:

In [50]:
my_numbers = [0,1,2,3,4,5,6,7,8,9]

Suppose that we wanted all the even numbers in the `my_numbers` list. We can use the double colon syntax to return every second element.

In [51]:
my_numbers[::2]

[0, 2, 4, 6, 8]

What about if we wanted all the odd numbers? Well this is where the magic of this syntax comes in. Phrased slightly more generally, we want every other element in this list starting from the first element. We know how to do each of these parts but not the whole problem. We use the following syntax to do this

In [52]:
my_numbers[1::2]

[1, 3, 5, 7, 9]

Okay cool. What about every even number starting from the element `2`? We can do something similar

In [53]:
my_numbers[2::2]

[2, 4, 6, 8]

How about every even number up 6?

In [54]:
my_numbers[:6:2]

[0, 2, 4]

How about every even number starting from 2 up to but not including 6?

In [55]:
my_numbers[2:6:2]

[2, 4]

Finally, just as an extra, if we want a quick way to reverse a list, we can use slicing to do so. We just use a negative step size.

In [57]:
my_numbers[::-1]

[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]

In [73]:
# Challenge: Return every odd number from 9 up to AND including 3 in descending order using the my_numbers list

[9, 7, 5, 3]

#### Important List Methods